Support for DynamodbEnhancedClientExtensions

knuspertante opened the pull request:
Refs: [DynamoDbEnhancedClient - Feature Request - Support Extensions like AutoGeneratedTimestampRecordExtension](https://github.com/quarkiverse/quarkus-amazon-services/issues/469)

closes #469 

This PR is based on https://github.com/quarkiverse/quarkus-amazon-services/pull/473+ and could be merged afterwards

knuspertante commented:
@gsmet if you have time could you take a look at this? Any tips or suggestions are welcome.

gsmet commented on dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamoDbEnhancedBuildTimeConfig.java:
```diff
@@ -0,0 +1,29 @@
+     */
+    @ConfigItem
+    @ConfigDocSection
+    Optional<List<DynamodbEnhancedClientExtension>> clientExtensions = Optional
+            .of(List.of(DynamodbEnhancedClientExtension.VersionedRecordExtension,
+                    DynamodbEnhancedClientExtension.AutoGeneratedTimestampRecordExtension));
```
I'm not sure we should enable the extensions by default. Also that wouldn't be the way to do it.

knuspertante replied:
You are right! I removed this.

gsmet commented on dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamodbEnhancedClientRecorder.java:
```diff
@@ -28,10 +43,32 @@ public BeanContainerListener setDynamoDbAsyncClient() {
+
+    private List<DynamoDbEnhancedClientExtension> createExtensionList() {
+        if (this.buildTimeConfig.clientExtensions.isPresent() && !this.buildTimeConfig.clientExtensions.get().isEmpty()) {
+            return this.buildTimeConfig.clientExtensions.get().stream().map(configValue -> {
+                DynamoDbEnhancedClientExtension extension = ChainExtension.create();
+                switch (configValue) {
+                    case AtomicCounterExtension:
+                        extension = AtomicCounterExtension.builder().build();
+                        break;
+                    case VersionedRecordExtension:
+                        extension = VersionedRecordExtension.builder().build();
+                        break;
+                    case AutoGeneratedTimestampRecordExtension:
+                        extension = AutoGeneratedTimestampRecordExtension.create();
+                        break;
+                }
+                return extension;
+            }).collect(Collectors.toList());
```
We try to avoid lambdas in runtime code because they have a cost.

knuspertante replied:
Done.

gsmet reviewed (changes requested):
Thanks for this work. I added a few questions. Let's discuss things before you pushing a new version.

/cc @scrocquesel @hamburml also for additional opinions.

knuspertante commented on dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamoDbEnhancedBuildTimeConfig.java:
```diff
@@ -0,0 +1,29 @@
+            .of(List.of(DynamodbEnhancedClientExtension.VersionedRecordExtension,
+                    DynamodbEnhancedClientExtension.AutoGeneratedTimestampRecordExtension));
+
+    public enum DynamodbEnhancedClientExtension {
+        VersionedRecordExtension,
+        AtomicCounterExtension,
+        AutoGeneratedTimestampRecordExtension
+
+    }
```
Ok, I understand your thoughts. 

I think per config we should only support extensions provided by aws-sdk.
 
For custom extensions a simple producer would work? 

Or do you have a idea, how we could support custom extensions also?

scrocquesel replied:
To avoid the closed enum, maybe support extensions like quarkus cxf support features in application.properties ?

https://quarkiverse.github.io/quarkiverse-docs/quarkus-cxf/dev/user-guide/cxf-interceptors-and-features.html
https://github.com/quarkiverse/quarkus-cxf/blob/b23ec118730538d2f6532e1df786de39d684f2cd/extensions/core/runtime/src/main/java/io/quarkiverse/cxf/transport/CxfHandler.java#L120

It first try to load a named bean and if not available it instanciates the class.

The configuration would look like

````
quarkus.dynamodbenhanced.client-extensions=software.amazon.awssdk.enhanced.dynamodb.extensions.VersionedRecordExtension,MyCustomNamedBeanProvideByAProducer
````

scrocquesel replied:
you may also look at how interceptors are handled here

hamburml replied:
I think it is probably easier to support the extensions provides by the sdk at first. But I really like the idea from @scrocquesel to config this via properties which also enables custom extensions. If the effort is too high for the beginning we should stick to the provided extensions.

scrocquesel replied:
I played a bit to find something working https://github.com/knuspertante/quarkus-amazon-services/pull/1.
At least, I would keep the fully qualified name for the config.

I also add the generation of the documentation of the config items.

knuspertante replied:
@scrocquesel thx for your work! I merged your [PR#1](https://github.com/knuspertante/quarkus-amazon-services/pull/1)

scrocquesel replied:
Could you check that it works in native ? If not, we may need to pass a collection of `RecorderContext recorderContext.classProxy(className)` to the recorder instead of the set of class names (or produce `ReflectiveClassBuildItem`). RecorderContext can be injected in the build step. See https://quarkus.io/guides/writing-extensions#service-files

knuspertante replied:
Sure. I run the tests in native-mode. So far, all works great ;-) Thanks for your work!

gsmet replied:
I think we should go with @scrocquesel suggestion and provide shortcurts for well know classes.

You could have `versioned-record`, `atomic-counter` and `auto-generated-timestamp-record` as shortcuts that would be automatically transformed to the SDK classes and you could also push your own classes.

That way you could have the best of both worlds. The shortcuts would be in the javadoc so properly documented.

scrocquesel replied:
I push a fix to @knuspertante repo. This was not working for custom extension.

@gsmet This is not what have been done for interceptors, and we will need to support the mapping on the long term.

From the integration tests, I understood that such extensions works with annotations on table entity. Maybe we can plan to discover needed extensions and add them from the model if the property is not set ?

gsmet replied:
> @gsmet This is not what have been done for interceptors, and we will need to support the mapping on the long term.

For the former, I don't know. For the latter, probably better that we maintain the mapping centrally rather than requiring people to update their applications if they change.
But I won't fight on this, feel free to ignore me :).

knuspertante replied:
My personal opinion: the current status is the most consistent way, since only extensions of the aws-sdk could be currently recognized via annotations. Also all extensions are specified with full classnames.

knuspertante commented on dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamodbEnhancedClientRecorder.java:
```diff
@@ -28,10 +43,32 @@ public BeanContainerListener setDynamoDbAsyncClient() {
+                return extension;
+            }).collect(Collectors.toList());
+        }
+        return List.of(VersionedRecordExtension.builder().build());
```
VersionedRecordExtension is the only default Extension from aws-sdk. 

See: https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/enhanced/dynamodb/internal/client/ExtensionResolver.html#defaultExtensions()

I changed this line to: 

```java
return ExtensionResolver.defaultExtensions();
```

hamburml replied:
That is interesting. This PR adds the support for EnhancedClientExtensions, right? So right now there is no support and no ClientExtension is active, right? So what does happen when it is already used (in my case for example) and i update to the version which adds VersionedRecordExtension - will it break or not?

knuspertante replied:
@hamburml VersionRecordExtension is enabled per default, as long you did not defined others. 
Therefore I think all is fine and nothing will break. 

In order to load all default extensions in the future, I have now changed this line in my last [commit](https://github.com/quarkiverse/quarkus-amazon-services/pull/488/commits/a990941ba6ff43b3d136dcbc21f549b0f42fa0c1) to 

```java
return ExtensionResolver.defaultExtensions();
```

gsmet reviewed (changes requested):
Just a note that we need to clean up the history before merging.

scrocquesel commented on dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamodbEnhancedClientRecorder.java:
```diff
@@ -1,21 +1,40 @@
             public void created(BeanContainer container) {
                 DynamodbEnhancedClientProducer producer = container.beanInstance(DynamodbEnhancedClientProducer.class);
-                producer.setDynamoDbClient(container.beanInstance(DynamoDbClient.class));
+                producer.setDynamoDbClient(container.beanInstance(DynamoDbClient.class), createExtensionList());
```
I would createExtensionList sooner to fail faster in case of misconfiguration in the recorder method and pass the value as closure to the listener. Or if we want to share the same instance for both sync and async, we can have the createExtensionList return a `RuntimeValue<DynamoDbEnhancedClientExtension>` and have it inject from the processor.

In recorder, something like
```
return  new RuntimeValue<>(ExtensionResolver.resolveExtensions(extensions));
````

and in the processor

```
var extensions = recorder.createExtensionList();

containerListenerProducer.produce(new BeanContainerListenerBuildItem(recorder.setDynamoDbClient(extensions )));
```

knuspertante replied:
Great! Thanks for this advice! I took your tips and implement it.

scrocquesel reviewed (changes requested):
Some last changes.
Also, can you add a configuration validation in the processor like https://github.com/quarkiverse/quarkus-amazon-services/blob/31c63f541c1eb55528e7ca4cf8bcf1d9061071c7/common/deployment/src/main/java/io/quarkus/amazon/common/deployment/AmazonServicesClientsProcessor.java#L69-L89

knuspertante commented on dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamodbEnhancedClientRecorder.java:
```diff
@@ -28,10 +47,49 @@ public BeanContainerListener setDynamoDbAsyncClient() {
+    private List<DynamoDbEnhancedClientExtension> createExtensionList() {
+
+        List<DynamoDbEnhancedClientExtension> extensions = new ArrayList<>();
+        for (String item : buildTimeConfig.clientExtensions.orElse(Collections.emptyList())) {
```
As gsmet wrote earlier, they try to avoid using lambdas/streams here, cause of performance. Therefore I changed that to for-each. I think this is also easy to read and understand.

hamburml commented:
I don't like that the current integrationtest ist reused and the extension is added there. Isn't it cleaner to add another integrationtest which only tests the DynamoDbEnhancedClientExtension?

We could have the following structure:
- integrationtest without any extensions with blocking and async client (application.properties do not set quarkus.dynamodbenhanced.client-extensions).
- integrationtest with default extensions and own extensions with blocking and async client (application.properties with quarkus.dynamodbenhacned.client-extensions).

scrocquesel commented on dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamodbEnhancedClientRecorder.java:
```diff
@@ -1,37 +1,96 @@
+        if (extensions.isEmpty()) {
+            extensions.addAll(ExtensionResolver.defaultExtensions());
+        }
+        return new RuntimeValue<>(extensions);
```
You can use ExtensionResolver.resolveExtension to reduce the list to one Extension and get rid of the list.

knuspertante replied:
Thanks, no I got it!

hamburml commented on integration-tests/src/main/java/io/quarkus/it/amazon/dynamodb/enhanced/DynamoDbEnhancedResource.java:
```diff
@@ -88,4 +91,40 @@ public String testBlockingDynamo() {
     }
+
+    @GET
+    @Path("blocking/extensions")
```
What is the difference between paths blocking/extensions and blocking? You mentioned that VersionRecordExtension is enabled by default so existingTableEntry.getCreatedAtTimestamp() should also work in testBlockingDynamo() shouldn't it? And where is blocking/extensions called in the tests?

scrocquesel replied:
> What is the difference between paths `blocking/extensions` and `blocking`? You mentioned that `VersionRecordExtension` is enabled by default so `existingTableEntry.getCreatedAtTimestamp()` should also work in `testBlockingDynamo()` shouldn't it? And where is blocking/extensions called in the tests?

I don't know dynamodb that much but I guess `VersionRecordExtension` increment a counter property annotated with a associated annotation while `AutoGeneratedTimestampRecordExtension` is the one that actually set DateTime properties.

hamburml replied:
Oh, didn't see that `AutoGeneratedTimestampRecordExtension` is set in application.properties. Okay, got it thanks. Nevertheless in [AmazonDynamoDBEnhancedTest.java](https://github.com/quarkiverse/quarkus-amazon-services/pull/488/files#diff-7f09dd446cd33221d3ee77864353414406dfff169043e9c2e66a1b6282415b78) I only see calls to /async and /blocking but no /blocking/extensions. Maybe I am missing something, it's late here.

scrocquesel commented:
> I am not fond of reusing the current integrationtest and add the extension there. Isn't it cleaner to add another integrationtest which only tests the DynamoDbEnhancedClientExtension?
> 
> We could have the following structure:
> 
> * integrationtest without any extensions with blocking and async client (application.properties without quarkus.dynamodbenhanced.client-extensions). We need to make sure that everything still works without setting this property.
> * integrationtest with default extensions and own extensions with blocking and async client (application.properties with quarkus.dynamodbenhacned.client-extensions).

+1, would you mind adding the TestProfile to dynamiccally set the value of the properties ?

> I am not fond of reusing the current integrationtest and add the extension there. Isn't it cleaner to add another integrationtest which only tests the DynamoDbEnhancedClientExtension?
> 
> We could have the following structure:
> 
> * integrationtest without any extensions with blocking and async client (application.properties without quarkus.dynamodbenhanced.client-extensions). We need to make sure that everything still works without setting this property.
> * integrationtest with default extensions and own extensions with blocking and async client (application.properties with quarkus.dynamodbenhacned.client-extensions).

+1 would you mind adding the two QuarkusTestProfile ? WithDefaultExtensions and WithCustomExtensions ?

knuspertante commented:
> I am not fond of reusing the current integrationtest and add the extension there. Isn't it cleaner to add another integrationtest which only tests the DynamoDbEnhancedClientExtension?
> 
> We could have the following structure:
> 
> integrationtest without any extensions with blocking and async client (application.properties without quarkus.dynamodbenhanced.client-extensions). We need to make sure that everything still works without setting this property.
integrationtest with default extensions and own extensions with blocking and async client (application.properties with quarkus.dynamodbenhacned.client-extensions).

Sure, this makes sense! 

@scrocquesel @hamburml 

Just to be clear, default extensions are activate per default until you provide one explicit extension. 

When you do not set `quarkus.dynamodbenhacned.client-extensions` in application.properties **only** `software.amazon.awssdk.enhanced.dynamodb.extensions.VersionedRecordExtension` will be activated (just like before).

If you set `quarkus.dynamodbenhacned.client-extensions` in application.properties and want to use `software.amazon.awssdk.enhanced.dynamodb.extensions.VersionedRecordExtension` and some others like `software.amazon.awssdk.enhanced.dynamodb.extensions.AutoGeneratedTimestampRecordExtension` or any custom extension you need to set it explicit. 

Like: `quarkus.dynamodbenhanced.client-extensions=software.amazon.awssdk.enhanced.dynamodb.extensions.VersionedRecordExtension,io.quarkus.it.amazon.dynamodb.enhanced.MyExtension`

Now I implement a QuarkusTestProfile for custom extension, which only test our custom extension. Should I add a test for `software.amazon.awssdk.enhanced.dynamodb.extensions.AutoGeneratedTimestampRecordExtension` too?

scrocquesel commented:
> > I am not fond of reusing the current integrationtest and add the extension there. Isn't it cleaner to add another integrationtest which only tests the DynamoDbEnhancedClientExtension?
> > 
> > We could have the following structure:
> > 
> > integrationtest without any extensions with blocking and async client (application.properties without quarkus.dynamodbenhanced.client-extensions). We need to make sure that everything still works without setting this property.
> integrationtest with default extensions and own extensions with blocking and async client (application.properties with quarkus.dynamodbenhacned.client-extensions).
> 
> Sure, this makes sense! 
> 
> @scrocquesel @hamburml 
> 
> Just to be clear, default extensions are activate per default until you provide one explicit extension. 
> 
> When you do not set `quarkus.dynamodbenhacned.client-extensions` in application.properties **only** `software.amazon.awssdk.enhanced.dynamodb.extensions.VersionedRecordExtension` will be activated (just like before).
> 
> If you set `quarkus.dynamodbenhacned.client-extensions` in application.properties and want to use `software.amazon.awssdk.enhanced.dynamodb.extensions.VersionedRecordExtension` and some others like `software.amazon.awssdk.enhanced.dynamodb.extensions.AutoGeneratedTimestampRecordExtension` or any custom extension you need to set it explicit. 
> 
> Like: `quarkus.dynamodbenhanced.client-extensions=software.amazon.awssdk.enhanced.dynamodb.extensions.VersionedRecordExtension,io.quarkus.it.amazon.dynamodb.enhanced.MyExtension`
> 
> Now I implement a QuarkusTestProfile for custom extension, which only test our custom extension. Should I add a test for `software.amazon.awssdk.enhanced.dynamodb.extensions.AutoGeneratedTimestampRecordExtension` too?

Test all extensions at once. We only need to check that extensions are configured and not testing their behavior individually.

knuspertante commented:
> > > I am not fond of reusing the current integrationtest and add the extension there. Isn't it cleaner to add another integrationtest which only tests the DynamoDbEnhancedClientExtension?
> > > We could have the following structure:
> > > integrationtest without any extensions with blocking and async client (application.properties without quarkus.dynamodbenhanced.client-extensions). We need to make sure that everything still works without setting this property.
> > > integrationtest with default extensions and own extensions with blocking and async client (application.properties with quarkus.dynamodbenhacned.client-extensions).
> > 
> > 
> > Sure, this makes sense!
> > @scrocquesel @hamburml
> > Just to be clear, default extensions are activate per default until you provide one explicit extension.
> > When you do not set `quarkus.dynamodbenhacned.client-extensions` in application.properties **only** `software.amazon.awssdk.enhanced.dynamodb.extensions.VersionedRecordExtension` will be activated (just like before).
> > If you set `quarkus.dynamodbenhacned.client-extensions` in application.properties and want to use `software.amazon.awssdk.enhanced.dynamodb.extensions.VersionedRecordExtension` and some others like `software.amazon.awssdk.enhanced.dynamodb.extensions.AutoGeneratedTimestampRecordExtension` or any custom extension you need to set it explicit.
> > Like: `quarkus.dynamodbenhanced.client-extensions=software.amazon.awssdk.enhanced.dynamodb.extensions.VersionedRecordExtension,io.quarkus.it.amazon.dynamodb.enhanced.MyExtension`
> > Now I implement a QuarkusTestProfile for custom extension, which only test our custom extension. Should I add a test for `software.amazon.awssdk.enhanced.dynamodb.extensions.AutoGeneratedTimestampRecordExtension` too?
> 
> Test all extensions at once. We only need to check that extensions are configured and not testing their behavior individually.

Ok, then all should be fine now. 

No test required for `software.amazon.awssdk.enhanced.dynamodb.extensions.AutoGeneratedTimestampRecordExtension`, cause this extension is not configured.

scrocquesel commented on integration-tests/src/main/java/io/quarkus/it/amazon/dynamodb/enhanced/DynamoDbEnhancedResource.java:
```diff
@@ -13,6 +13,7 @@
 import org.jboss.logging.Logger;
 
 import software.amazon.awssdk.enhanced.dynamodb.*;
+import software.amazon.awssdk.services.dynamodb.model.ResourceInUseException;
```
you should be able to revert this file and also the application.properties

knuspertante commented:
@gsmet how does it looks to you? If I have to change anything, pls give me some feedback. Thanks!

scrocquesel commented:
> @gsmet how does it looks to you? If I have to change anything, pls give me some feedback. Thanks!

LGTM

gsmet reviewed (approved):
I amended your commit to remove the star imports that we try to avoid in Quarkus codebase.

Other than that, let's merge once CI is green (again).

Do you need that soon and should I backport it to a 1.x version?

scrocquesel commented:
> I amended your commit to remove the star imports that we try to avoid in Quarkus codebase.
> 

@gsmet is it something that could be enforced from the quarkiverse parent pom like the formatter ?

knuspertante commented:
Oh, the build / test fails, cause we use a Quarkus test-profil and the native-image is build with a different quarkus-profile. 

See: 
```shell
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2023-03-02 09:54:55,374 INFO  [io.net.uti.int.PlatformDependent] (Thread-1) Your platform does not provide complete low-level API for accessing direct buffers reliably. Unless explicitly requested, heap buffer will always be preferred to avoid potential system instability.
2023-03-02 09:54:55,406 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.kinesis.aws.region" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2023-03-02 09:54:55,406 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.kinesis.aws.credentials.type" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2023-03-02 09:54:55,407 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.redshift.aws.credentials.type" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2023-03-02 09:54:55,407 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.kinesis.aws.credentials.static-provider.secret-access-key" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2023-03-02 09:54:55,407 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.redshift.endpoint-override" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2023-03-02 09:54:55,408 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.kinesis.endpoint-override" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2023-03-02 09:54:55,408 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.redshift.aws.region" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2023-03-02 09:54:55,409 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.redshift.aws.credentials.static-provider.secret-access-key" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2023-03-02 09:54:55,409 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.kinesis.aws.credentials.static-provider.access-key-id" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2023-03-02 09:54:55,410 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.redshift.aws.credentials.static-provider.access-key-id" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2023-03-02 09:54:55,412 WARN  [io.qua.run.con.ConfigRecorder] (main) The profile 'prod' used to build the native image is different from the runtime profile 'test-dynamodb-enhanced-client-with-custom-extensions'. This may lead to unexpected results.
2023-03-02 09:54:55,422 INFO  [io.quarkus] (main) quarkus-amazon-services-integration-tests 999-SNAPSHOT native (powered by Quarkus 3.0.0.Alpha4) started in 0.056s. Listening on: http://0.0.0.0:8081/
2023-03-02 09:54:55,422 INFO  [io.quarkus] (main) Profile prod activated. 
2023-03-02 09:54:55,423 INFO  [io.quarkus] (main) Installed features: [amazon-cognito-user-pools, amazon-dynamodb, amazon-dynamodb-enhanced, amazon-iam, amazon-kms, amazon-s3, amazon-secretsmanager, amazon-ses, amazon-sns, amazon-sqs, amazon-ssm, cdi, resteasy-reactive, smallrye-context-propagation, vertx]
2023-03-02 09:54:55,512 INFO  [io.qua.it.ama.dyn.enh.DynamoDbEnhancedResource] (executor-thread-0) Testing Blocking Dynamodb client with table: enhancedblocking
Error:  Tests run: 2, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 35.838 s <<< FAILURE! - in io.quarkus.it.amazon.AmazonDynamoDBEnhancedWithCustomExtensionITCase
Error:  testDynamoDbEnhancedClientWithCustomExtensionAsync  Time elapsed: 4.928 s  <<< FAILURE!
java.lang.AssertionError: 
1 expectation failed.
Response body doesn't match expectation.
Expected: is "INTERCEPTED EXTENSION OK@1"
  Actual: INTERCEPTED OK@1


Error:  testDynamoDbEnhancedClientWithCustomExtensionBlocking  Time elapsed: 5.2 s  <<< FAILURE!
java.lang.AssertionError: 
1 expectation failed.
Response body doesn't match expectation.
Expected: is "INTERCEPTED EXTENSION OK@1"
  Actual: INTERCEPTED OK@1
```

I think this is already discussed here: https://github.com/quarkusio/quarkus/issues/4371

Do you have any suggestions? 

Currently I used a producer as workaround but some others are waiting for this. See: https://github.com/quarkiverse/quarkus-amazon-services/issues/469#issuecomment-1446177127

scrocquesel commented:
I would make the profile with extensions the default to be sure native build don't mess up with bean. And remove the @QuarkusIntegrationTest annotated class for the case without extensions.

knuspertante commented:
@scrocquesel can you pls have a look on my last commit. Native-Tests works, now all should be fine. Thanks for your advice!

scrocquesel commented:
> @scrocquesel can you pls have a look on my last commit. Native-Tests works, now all should be fine. Thanks for your advice!

@knuspertante it seems you remove both `@QuarkusIntegrationTest` class. Could you restore

```java
@QuarkusIntegrationTest
public class AmazonDynamoDBEnhancedITCase extends AmazonDynamoDBEnhancedTest { 
}
```

knuspertante commented:
@scrocquesel no I left one ;-) https://github.com/knuspertante/quarkus-amazon-services/blob/8792e1a03462eca2c4302f3569c7853f11cdd5dd/integration-tests/src/test/java/io/quarkus/it/amazon/AmazonDynamoDBEnhancedITCase.java

I made the withExtensionTest the default case. So I switched the classes and renamed the test-profile and the test-class to WithoutExtension

scrocquesel commented:
@gsmet  /lgtm + backport 1.x and release 1.x please

scrocquesel commented:
> @gsmet /lgtm + backport 1.x and release 1.x please

@gsmet kindly reminder

gsmet commented:
Sorry, I was extremely busy on preparing Quarkus 3. Going through the PRs now.

knuspertante commented:
> Sorry, I was extremely busy on preparing Quarkus 3. Going through the PRs now.

No problem, thanks for merging! How is your plan to include it in future releases or backport it as suggested by scrocquesel?

gsmet commented:
I already backported it and will push new versions for both Quarkus 2 and 3 soon.

gsmet commented:
@knuspertante 1.7.0 is on its way to Central and release of 2.0.0 is in progress.
